rfcs/20190502-preprocessing-layers.md

```diff
@@ -0,0 +1,578 @@
+```
+
+### Writing a subclass of `PreprocessingLayer`
+
+The following 3 methods should be overridden:
+
+- `__init__`: constructor of the layer, used to configure its behavior.
+- `build(self, inputs_shape)`: creates the state variables of the layer.
+- `call(self, inputs)`: transforms the inputs (should only be called after `adapt` has been called).
+- `adapt(self, data, [reset_state=True])`: sets the state of the layer given the data provided (either as a tf.data dataset or numpy array(s)). The `reset_state` argument is optional and may be ignored.
+
+
+### Handling of async prefetching
+
+Some preprocessing ops are CPU-only and benefit from being executed asynchronously on the accelerator host (as opposed to the accelerator itself, e.g. GPU or TPU),
+with a batch of data being prepocessed on the host while the previous batch is being processed by the accelerator. This pattern is known as "async prefetching".
+
+This is normally done as part of a tf.data pipeline. The current proposal implies moving some of that preprocessing to inside the model itself, which is normally
+executed end-to-end on an accelerator.
+
+This means that we need a way to lift the preprocessing part of the model in a tf.data pipeline during model training. In `fit`, we can do this automatically.
+In custom training loops, we will expect the user to do it manually.
+
+We propose the addition of two new methods on the `Model` class:
+
+```python
+def get_preprocessing_stage(self):
+    """Retrieves the preprocessing part of the model.
+
+    This is the part of the model that should be executed asynchronously
+    on the device host during training.
+
+    Returns:
+        Instance of `PreprocessingLayer` or `PreprocessingStage`.
+        May be None if the model does not start with preprocessing layers.
+    """
+    pass
+
+def get_main_stage(self):
+    """Retrieves the main processing part of the model.
+
+    This is the part of the model that should be executed
+    on the accelator device.
+
+    Returns:
+        Model instance.
+    """
+```
+
+Thus, for any model that starts with preprocessing layers, the following:
+
+```python
+outputs = model(inputs)
+```
+
+is functionnally equivalent to:
+
+```python
+preprocessed_inputs = model.get_preprocessing_stage()(inputs)
+outputs = model.get_main_stage()(preprocessed_inputs)
+```
+
+
+#### Examples:
+
+Sequential model with a preprocessing layer:
+
+```python
+vectorization = keras.layers.TextVectorization()
+vectorization.adapt(data_sample)
+
+model = keras.Sequential([
+    vectorization,
+    keras.layers.Dense(10, activation='softmax'),
+])
+
+# This is the `vectorization` layer.
+preproc_stage = model.get_preprocessing_stage()
+# model containing the `Dense` layer only.
+main_stage = model.get_main_stage()
+```
+
+Functional model with 2 branches, each with a preprocessing layer:
+
+```python
+normalization_a = layers.Normalization()
+normalization_b = layers.Normalization()
+normalization_a.adapt(data_a)
```

Comment:
I guess this would also support Dataset, Sequence, etc like fit_generator?

Reply:
Ideally yes, but since this method has to be overridden by layer implementers, there are complications.

Reply:
Does this mean that the whole `data` should be in memory first?

Reply:
Not necessarily, it is passed as a dataset, then the `adapt` method will iterate on the dataset in batches.